### ISSUE TITLE
DCC GET sometimes hangs due to ack

### ISSUE THREAD

**Issue by martinetd** [weechat/weechat#1167]
Sometimes while getting a file the transfer would just stop progressing and end up timeouting and failing.
I finally managed to strace the child process while that happens and here's what I can see:
```
sendto(38, "\372R", 2, 0, NULL, 0)      = -1 EAGAIN (Resource temporarily unavailable)
nanosleep({tv_sec=0, tv_nsec=1000000}, NULL) = 0
sendto(38, "\372R", 2, 0, NULL, 0)      = -1 EAGAIN (Resource temporarily unavailable)
nanosleep({tv_sec=0, tv_nsec=1000000}, NULL) = 0
sendto(38, "\372R", 2, 0, NULL, 0)      = -1 EAGAIN (Resource temporarily unavailable)
nanosleep({tv_sec=0, tv_nsec=1000000}, NULL) = 0
sendto(38, "\372R", 2, 0, NULL, 0)      = -1 EAGAIN (Resource temporarily unavailable)
nanosleep({tv_sec=0, tv_nsec=1000000}, NULL) = 0
sendto(38, "\372R", 2, 0, NULL, 0)      = -1 EAGAIN (Resource temporarily unavailable)
nanosleep({tv_sec=0, tv_nsec=1000000}, NULL) = 0
```

basically, sendto() of only 2 means we've sent two bytes of the ack so we're retrying continuously in the loop here:
```c
int
xfer_dcc_recv_file_send_ack (struct t_xfer *xfer)
{
    int length, num_sent, total_sent;
    uint32_t pos;
    const void *ptr_buf;

    pos = htonl (xfer->pos);
    ptr_buf = &pos;
    length = 4;
    total_sent = 0;
    num_sent = send (xfer->sock, ptr_buf, length, 0);
    if (num_sent > 0)
        total_sent += num_sent;
    while (total_sent < length)
    {
        if ((num_sent == -1) && (errno != EAGAIN) && (errno != EWOULDBLOCK))
            return 0;

        /* if we can't send ACK now, just return with partial failure code */
        if (total_sent == 0)
            return 1;

        /* at least one byte has been sent, we must send whole ACK */
        usleep (1000);
        num_sent = send (xfer->sock, ptr_buf + total_sent,
                         length - total_sent, 0);
        if (num_sent > 0)
            total_sent += num_sent;
    }

    /* ACK successfully sent */
    return 2;
}
```

The problem being, it spends at least 180s (timeout value) doing that and never stops until the xfer aborts.
I suspect this is something silly like a shared buffer for send and recv that won't ever be available again until we've read some from the socket.

I don't see any good solution to this, reimplementing the read side in ack seems silly (it could read the last bytes of the transfer!), giving up here will cause trouble because we've not sent the whole int, the best I can come up with is returning some new values say `-n` meaning "there are n bytes left to send" and on next iteration only try to send the last n bytes of the int; but given network byte order is big endian we could send a value that is bigger than we intended to send (first iteration on 255 then second iteration on 256 will send `0x01ff` which is higher than both `0x0100` and `0x000ff`) so we'd need to remember the previous pos to finish from in that case.... Which isn't impossible but quite ugly.
An alternative might be to set TCP_NODELAY when creating the socket so the acks don't pile up and we'll always have room for either 0 or 4 bytes - but this feels very hackish.

To be blunt I don't understand this ack at all, this protocol doesn't make sense to me, and would be happy with an option to only send it at the end of the transfer...

Anyway, happy to provide with patch & test once a way out has been figured. If nothing happens I'll send a patch adding an option to turn off acks by default.

**System Info**
Weechat 2.0.1 (archlinux version)
linux 4.14.11-1-ARCH

**Comment by martinetd** [comment]
Note: I wrote that issue a bit early, and my "happy to test" might be limited by the fact that I'm moving out of the current place where I reproduce this behaviour once every ~5 big transfers at the end of the week.

To offset that a bit, I actually went ahead and added an `xfer.network.send_ack` option which I am currently testing. If it looks like it works (which I don't see why it wouldn't), I'll submit it in a PR at the end of the week keeping the default option as it is - but it's a very bad fix: folks with the problem will have no idea the option can help them.

I had started writing code that remembers how many bytes were written and would try again later but that gets very messy around the final ack, ultimately I think if the option works well and has no visible drawback changing the default value to not send ack on file receive by default might make more sense.

Until then, how about adding a counter in the retry loop and if we haven't sent any byte in, say, 10 seconds (10'000 iterations) then display a message in the core weechat buffer suggesting to turn the option on?

**Comment by flashcode** [comment]
Hi @martinetd,

Maybe enabling the option TCP_NODELAY (only on the receiver side, which is sending ACK) would solve the problem?
Since the receiver only sends ACK (so very small packets), and that these ACKs must be sent immediately, this option should do its job. Is there any drawback to enable this option?

**Comment by martinetd** [comment]
Hi!

Currently the acks are piled up and sent in batches (in pcap files I see anywhere from 1 ack alone to mtu-sized packets full of acks, but it's mostly around ~300 bytes long so ~75 acks at a time in my use case).

Using TCP_NODELAY would certainly change that so I'd say it could work, but I don't see any server actually waiting for these (ok, weechat without xfer.network.fast_send would, but it's on by default; irssi doesn't, iroffer doesn't), so I'm not sure about sending 4-bytes packets all the time, networks really don't like small packets... Only the final ack seems to actually be useful, but even here just closing the connection likely has a similar effect in practice.

I just don't understand this protocol, trying to reimplement TCP acks over TCP?!
I couldn't even find a reference/specification of the protocol when I looked it up earlier, and mirc seems to send 64bit acks instead of 32bit acks so it's not even clearly cut...

I will try with an extra `setsockopt` call with `TCP_NODELAY instead of my disabling for the rest of the week but I'd personally rather just disable acks altogether even if it's more aggressive, at least until I see an example of server that'd really want them.

**Comment by martinetd** [comment]
hmm, interestingly, even setting `TCP_NODELAY` doesn't make all acks be sent one at a time, but it's certainly sent out more often; the bulk acks that were sent out every ~40 TCP ack packets are now sent 1 to 15 at a time every 2-4 TCP acks.

There still are occasional big packets (I saw one with 1068 bytes, 267 acks) so I'm not sure it'll fix my problem, it might make it less frequent but that feels like pushing it under the rug.

I also got 2 of 3 transfers just now hit the path that disable acks naturally (`xfer_dcc_recv_file_send_ack` returning 1 after 0 bytes were sent), I *think* this might happen more often that it did without `TCP_NODELAY`, but once again since there doesn't seem to be any downside of no longer sending acks I'd rather just not send them than see the weechat logic somtimes disable it in the middle of the transfer or not from a debugging point of view. If I ever run into a server that waits for them it'll be very hard to figure out...
